### ISSUE TITLE
RecursionError: maximum recursion depth exceeded

### ISSUE THREAD

**Issue by anugrahsoft** [django-tenants/django-tenants#949]
Hi, 

I installed a new Debian box (version 12), wich comes along with PostgreSQL 15 and Python 3.11.2. I also iinstalled the latest Django 4.2.2

Just to let you know that every other django project works fine, except the one that I'm using django-tenants.

I got this error : "RecursionError: maximum recursion depth exceeded" , whenever I execute : python manage.py makemigrations

Before on Debian 11 : PostgreSQL13, Python 3.9, Django 4.2, everythin works great.

I don't know, but I do suspect django-tenants, because on the installation of my -r requirements.txt, it says this warning :
```
(venv) rapsody@znb:~/dev/healmaster_project/healmaster$ pip install django-tenants
Collecting django-tenants
  Using cached django-tenants-3.5.0.tar.gz (117 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: Django<4.3,>=2.1 in ./venv/lib/python3.11/site-packages (from django-tenants) (4.2.1)
Requirement already satisfied: asgiref<4,>=3.6.0 in ./venv/lib/python3.11/site-packages (from Django<4.3,>=2.1->django-tenants) (3.7.2)
Requirement already satisfied: sqlparse>=0.3.1 in ./venv/lib/python3.11/site-packages (from Django<4.3,>=2.1->django-tenants) (0.4.4)
Installing collected packages: django-tenants
  DEPRECATION: django-tenants is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  Running setup.py install for django-tenants ... done
Successfully installed django-tenants-3.5.0
(venv) rapsody@znb:~/dev/healmaster_project/healmaster$ 
```

here's the trace for the error, casued by the command : python manage.py makemigrations:
```
(venv) rapsody@znb:~/dev/healmaster_project/healmaster$ python manage.py makemigrations
 
>>> manage.py <<<
>>> Local development <<<
Traceback (most recent call last):
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 114, in debug_sql
    yield
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 102, in execute
    return super().execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 87, in _execute
    return self.cursor.execute(sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/cursor.py", line 719, in execute
    self._conn.wait(
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/connection.py", line 957, in wait
    return waiting.wait(gen, self.pgconn.socket, timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/waiting.py", line 262, in wait_epoll
    s = next(gen)
        ^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/cursor.py", line 195, in _execute_gen
    pgq = self._convert_query(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/client_cursor.py", line 78, in _convert_query
    pgq = PostgresClientQuery(self._tx)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/_queries.py", line 50, in __init__
    self._encoding = conn_encoding(transformer.connection)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/psycopg/_encodings.py", line 87, in conn_encoding
    if not conn or conn.closed:
                   ^^^^^^^^^^^
RecursionError: maximum recursion depth exceeded

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rapsody/dev/healmaster_project/healmaster/manage.py", line 82, in <module>
    execute_from_command_line(sys.argv)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/core/management/commands/makemigrations.py", line 156, in handle
    loader.check_consistent_history(connection)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/migrations/loader.py", line 313, in check_consistent_history
    applied = recorder.applied_migrations()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/migrations/recorder.py", line 81, in applied_migrations
    if self.has_table():
       ^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/migrations/recorder.py", line 57, in has_table
    with self.connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 144, in _cursor
    cursor = super()._cursor()
             ^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 306, in _cursor
    self.ensure_connection()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
    self.connect()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 272, in connect
    self.init_connection_state()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 309, in init_connection_state
    super().init_connection_state()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 239, in init_connection_state
    self.check_database_version_supported()
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 210, in check_database_version_supported
    and self.get_database_version() < self.features.minimum_database_version
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 188, in get_database_version
    return divmod(self.pg_version, 10000)
                  ^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
     






[[[[ ------> Text too long - Truncated -<------ ]]]]







  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 171, in _cursor
    cursor_for_search_path.execute('SET search_path = {0}'.format(','.join(formatted_search_paths)))
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 101, in execute
    with self.debug_sql(sql, params, use_last_executed_query=True):
  File "/usr/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/utils.py", line 119, in debug_sql
    sql = self.db.ops.last_executed_query(self.cursor, sql, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 312, in last_executed_query
    return self.compose_sql(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/operations.py", line 205, in compose_sql
    return mogrify(sql, params, self.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/psycopg_any.py", line 21, in mogrify
    with connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django_tenants/postgresql_backend/base.py", line 144, in _cursor
    cursor = super()._cursor()
             ^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 308, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
                                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rapsody/dev/healmaster_project/healmaster/venv/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 330, in create_cursor
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
RecursionError: maximum recursion depth exceeded
(venv) rapsody@znb:~/dev/healmaster_project/healmaster$
```

Thank you

**Comment by anugrahsoft** [comment]
Dowgrade it to psycopg2 worked.  

So Definetly there's something between django-tenants and psycopg3.

**Comment by godd0t** [comment]
Hello,

I've encountered a similar issue when using Django Tenants with psycopg3. I've created a new issue [Migrations not being applied to new tenant schemas with psycopg3 #950](https://github.com/django-tenants/django-tenants/issues/950) that describes the problem in detail.

In summary, when creating a new tenant and subsequently running the `migrate_schemas` command, the migrations are not being applied to the newly created tenant schema. This issue does not occur when using psycopg2.

I've found a workaround that involves modifying the `_cursor` method in the `DatabaseWrapper` class found in `django_tenants/postgresql_backend/base.py` to include a flag that prevents recursion. This seems to resolve the recursion issue, but the `migrate_schemas` command still does not apply the migrations to the newly created tenant schema when using psycopg3.

**Comment by anugrahsoft** [comment]
Hi godd0t,


Thank's for the tips, sorry I didn't see that you reported the issue, otherwise I wouldn't post mine...

indeed it was "postgresql_backend/base.py", since the Tom and the crew know already about this, we hope to see a hot fix soon,

to be honest my server was blocked by this issue on digital ocean, I upgrade the Debian distribution and I was stuck for a while until I troubleshoot things and come up with a quick way around to downgrade to psycopg2.

I do also notice some a great and slightly quickness using psycopg3 and Python 3.11, It was really smooth comparing to psycopg2 and Python 3.9.

I do also need "migrate_schemas" although I would've take your fix above,
If you know a quick fix that fix definitely the problem let me know, I'm a taker, until it will be officially fixed..

Thank you man.

**Comment by godd0t** [comment]
> Hi godd0t,
> 
> Thank's for the tips, sorry I didn't see that you reported the issue, otherwise I wouldn't post mine...
> 
> GOOD CATCH ! indeed it was "postgresql_backend/base.py", since Tom and the team knows already about this, we hope to see a hot fix soon,
> 
> To be honest my server was off by this issue on digital ocean, I upgrade the Debian distribution and I was stuck for a while until I troubleshoot things and came up with a quick way around to downgrade to psycopg2.
> 
> I do also noticed some great and slightly quickness using psycopg3 and Python 3.11, It was really smooth comparing to psycopg2 and Python 3.9.
> 
> I do also need "migrate_schemas" although I would've take your fix above, If you know a quick fix that fix definitely the problem let me know, I'm a taker, until it will be officially fixed..
> 
> Thank you man.

No need to apologize at all! Your post was helpful and it's always good to have more information and perspectives on these kinds of problems. We're all working together to improve the tools we use.

I'm glad to hear that you found a workaround by downgrading to psycopg2. I understand the appeal of using psycopg3 and Python 3.11, as I've also noticed the performance improvements.

As for the migrate_schemas issue, the workaround I posted does seem to prevent the recursion error, but it doesn't fully resolve the issue with migrations not being applied to the newly created tenant schema when using psycopg3. I'm still investigating this and will definitely share any updates or solutions I find.

**Comment by anugrahsoft** [comment]
Sure,

I would love to see your fix soon .

Thank you

**Comment by tomturner** [comment]
I will look at running DT with psycopg3 soon

**Comment by anugrahsoft** [comment]
Hi Tom,

Thank you so much for that. We Appreciate it.

**Comment by Jigsaw52** [comment]
I have looking into this and the issue was introduced with Django 4.2.1.
This commit is the cause of it:

https://github.com/django/django/commit/090d5ccc6c59e4df491ec0599c569834bdcf1409

**Comment by anugrahsoft** [comment]
Hi @Jigsaw52,

Thank's for bring that up, ... but I think that this issue is not related with the "RecursionError: maximum recursion depth exceeded
", but rather with another psycopg3 issue, wich is "Regression SearchHeadline" wich its ticket is "34486"

But I think this psycopg3 issue is related to new way of python tuple binding, wich its ticket is "34434" here :
https://code.djangoproject.com/ticket/34434

Wich has not been taken care of in django-teannts

I'm not sure though...

The ticket seems different !..

**Comment by Jigsaw52** [comment]
> Hi @Jigsaw52,
> 
> Thank's for bring that up, ... but I think that this issue is not related with the "RecursionError: maximum recursion depth exceeded ", but rather with another psycopg3 issue, wich is "Regression SearchHeadline" wich its ticket is "34486"
> 
> But I think this psycopg3 issue is related to new way of python tuple binding, wich its ticket is "34434" here : https://code.djangoproject.com/ticket/34434
> 
> Wich has not been taken care of in django-teannts
> 
> I'm not sure though...
> 
> The ticket seems different !..

The Django issue in the commit I mentioned (34486) is not related to psycopg3. However, it is the fix for that issue that is causing the infinite recursion error. Specifically, the addition of the following call to connection.cursor() causes it:

https://github.com/django/django/commit/090d5ccc6c59e4df491ec0599c569834bdcf1409#diff-01f6880f77beca32ee83e011072ba73dc7eed7f9f3efdebd935af693a4fac7b3R21

You verify that by reverting this Django commit, the infinite recursion problem no longer happens on Django 4.2.1.

**Comment by hho6643** [comment]
@anugrahsoft  I'm wondering if pull request #959 fixes the issue for you.

**Comment by Nasjoe** [comment]
Hi !
Same probleme here with 

django 4.2.7
django-tenant 3.5.0
psycopg 3.1.13

seem no problem when i downgrade to django 4.2
